Broader IPFS and IPNS URL support like http/s

markg85 opened the pull request:
Commit msg:
```
Previously just ipfs://<cid> and ipns://<cid> was supported. Which is too strict for some usecases.

This patch allows paths and query arguments to be used too. Making this work according to normal http semantics: ipfs://<cid>/foo/bar?key=val
ipns://<cid>/foo/bar?key=val

The gateway url support is changed.
It now only supports gateways in the form of:
http://<gateway>/foo/bar
http://<gateway>

Query arguments here are explicitly not allowed and trigger an intended malformed url error.

There also was a crash when IPFS_PATH was set with a non trailing forward slash. This has been fixed.

Lastly, a load of test cases have been added to verify the above.
```

This fixes https://github.com/curl/curl/issues/12148 and a couple more issues that had been found.

I've added a lot of testcases to verify the fixes.
Still, the testing isn't super elaborate (say everything for ipfs, then another time for ipns and then everything 3x for the 3 different ways to set a gateway). The reasoning is that much code is following the same path so an issue popping up for ipns will be there for ipfs too. As long as that stays as-is then those current tests cover a lot :)

cc @lidel @Stebalien

Stebalien commented on src/tool_operate.c:
```diff
@@ -697,45 +697,59 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
+  if(*input && **input) {
+    size_t len = strlen(*input);
+    if(((*input)[len - 1] != val_to_ensure)) {
+      char *composed_input = aprintf("%s%c", *input, val_to_ensure);
+      if(composed_input) {
+        Curl_safefree(*input);
+        *input = aprintf("%s", composed_input);
+        Curl_safefree(composed_input);
+      }
```
```suggestion
      char *composed_input = aprintf("%s%c", *input, val_to_ensure);
      Curl_safefree(*input);
      *input = composed_input;
```

(should be equivalent?)

Or even:

```c
      *input = Curl_saferealloc(*input, len+2); // needs some casting
      if(*input) {
         input[len] = val_to_ensure;
         input[len+1] = 0;
      }
```

But, I'm not sure if that violates some rule of curl's coding style.

markg85 replied:
I think that's fine, i see more of those constructions in curl.
Thank you for this hint!  This idea works just fine and saves a temp copy!

Stebalien commented on src/tool_operate.c:
```diff
@@ -751,10 +765,9 @@ static char *ipfs_gateway(void)
-      gateway = aprintf("%s%s", buf, (add_slash) ? "/" : "");
-    }
+    if((PARAM_OK == file2string(&buf, gateway_file)) && buf && *buf)
+      gateway = aprintf("%s", buf);
```
seems like a useless copy?

markg85 replied:
That's not entirely the case...

It gets the file content and checks it for non-null. If that is non-null than `gateway` is filled with it's value.
If the file would exist but is empty, this line would make `gateway` a non-null but empty variable.

That being said, i will restructure the checks to after the call which would clean it up nicely too.

Stebalien replied:
No, but copying `buf` to a new `gateway` buffer with `aprintf` is definitely a useless copy.

Stebalien reviewed (commented):
Basic feedback but I'm not familiar with curl's coding guidelines.

bagder commented on src/tool_operate.c:
```diff
@@ -697,45 +697,59 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
+      char *composed_input = aprintf("%s%c", *input, val_to_ensure);
+      if(composed_input) {
+        Curl_safefree(*input);
+        *input = aprintf("%s", composed_input);
```
Save a useless copy:

```suggestion
        *input = composed_input;
```

markg85 replied:
Fixed by restructuring it with the idea below from Stebalien.

bagder commented on src/tool_operate.c:
```diff
@@ -697,45 +697,59 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
+    size_t len = strlen(*input);
+    if(((*input)[len - 1] != val_to_ensure)) {
+      char *composed_input = aprintf("%s%c", *input, val_to_ensure);
+      if(composed_input) {
```
When this `aprintf()` call fails, shouldn't `*input` be freed and pointed to `""` or NULL or something?

markg85 replied:
That does mean every usage of this function needs to be checked `if(!ensure_trailing(..)) // error handling`.
I'll turn it into a function returning a CURLcode and checking for that on each invocation of this function.

bagder commented on src/tool_operate.c:
```diff
@@ -697,45 +697,59 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
-      gateway = aprintf("%s", composed_gateway);
-      Curl_safefree(composed_gateway);
-    }
+    ensure_trailing(&gateway, '/');
```
If `gateway` is NULL when this function returns, should it still return that NULL here or continue? And if it returns NULL here, should it not provide an error message to help the user why it stops?

markg85 commented:
@bagder 

```
[100%] Linking C executable curl
/usr/bin/ld: CMakeFiles/curl.dir/tool_operate.c.o: in function `ensure_trailing':
tool_operate.c:(.text+0x1384): undefined reference to `Curl_saferealloc'
collect2: error: ld returned 1 exit status
gmake[2]: *** [src/CMakeFiles/curl.dir/build.make:873: src/curl] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:526: src/CMakeFiles/curl.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
Error: Process completed with exit code 2.
```
Any idea why many compilers can't seem to find `Curl_saferealloc`? I did add the header for it `strdup.h`. Is this function exclusive to libcurl and can't be used in the tool?

For reference, on my local pc (gcc (GCC) 13.2.1 20230801) it works just fine.

bagder commented:
> I'll just use the normal realloc instead.

I recommend using a dynbuf instead. Because it is less easy to mess up.

bagder commented on src/tool_operate.c:
```diff
@@ -824,32 +878,95 @@ static CURLcode ipfs_url_rewrite(CURLU *uh, const char *protocol, char **url,
+  /* get the path from user input */
+  if(curl_url_get(uh, CURLUPART_PATH,
+                  &inputpath, CURLU_URLDECODE) != CURLUE_OK) {
+    inputpath = aprintf("%s", "");
```
```suggestion
    inputpath = strdup("");
    if(!inputpath)
       goto clean;
```

bagder commented on src/tool_operate.c:
```diff
@@ -824,32 +878,95 @@ static CURLcode ipfs_url_rewrite(CURLU *uh, const char *protocol, char **url,
+  if(inputpath && *inputpath && strlen(inputpath) == 1) {
+    if(*inputpath == '/') {
+      Curl_safefree(inputpath);
+      inputpath = aprintf("%s", "");
```
```suggestion
      inputpath = strdup("");
      if(!inputpath)
         goto clean;
```

bagder commented on tests/data/DISABLED:
```diff
@@ -70,9 +70,19 @@
 266
 579
 587
+# most of the IPFS tests, hyper doesn't like them
```
Why doesn't hyper like the IPFS? This seems like something we should fix, not hide.

markg85 replied:
It's the CRLF vs LF line endings. I just can't get those to work.
If you have more luck, please do. Last time i spend days trying to make hyper like it, nothing worked.

markg85 replied:
Hyper is fixed. I just needed to add `nocheck="yes"` to all tests that contain server data. I somehow completely missed that before. No IPFS tests are disabled for hyper now. Yay :)

bagder commented on src/tool_operate.c:
```diff
@@ -697,45 +697,75 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
+
+      Curl_safefree(*input);
+
+      if(curlx_dyn_addf(&dyn, "%c", val_to_ensure)) {
```
Leaner:

```suggestion
      if(curlx_dyn_addn(&dyn, &val_to_ensure, 1)) {      
```

bagder commented on src/tool_operate.c:
```diff
@@ -697,45 +697,75 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
+      struct curlx_dynbuf dyn;
+      curlx_dyn_init(&dyn, len + 2);
+
+      if(curlx_dyn_add(&dyn, *input)) {
```
Save a strlen:

```suggestion
      if(curlx_dyn_addn(&dyn, *input, len)) {
```

markg85 commented on src/tool_operate.c:
```diff
@@ -749,24 +779,36 @@ static char *ipfs_gateway(void)
+    while((c = getc(gateway_file)) != EOF && c != '\n' && c != '\r') {
+      if(curlx_dyn_addf(&dyn, "%c", (char) c)) {
+        /* reset the buffer, 0 length is going to trigger to cleanup below */
+        curlx_dyn_reset(&dyn);
```
Nice! Fixed.

bagder commented on src/tool_operate.c:
```diff
@@ -749,24 +778,33 @@ static char *ipfs_gateway(void)
+      ensure_trailing(&gateway, '/');
+
     if(!gateway) {
       Curl_safefree(gateway);
```
if gateway is already NULL, there is not much point in freeing it...

markg85 replied:
Nice! Removed.

markg85 commented:
Hi @bagder,

Just flattened it all to one commit.
It's probably ready to get merged now.

If it's ok with you, i'd prefer to wait merging till a second functional-review has been done by @lidel.
Is merging on Monday still fine to make it in the 8.5.0 release?

bagder commented:
> Is merging on Monday still fine to make it in the 8.5.0 release?

This is a bugfix, we can merge such up until December 6...

bagder commented on src/tool_operate.c:
```diff
@@ -697,45 +697,74 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
-      gateway = aprintf("%s", composed_gateway);
-      Curl_safefree(composed_gateway);
+  if(gateway) {
+    if(ensure_trailing(&gateway, '/') != CURLE_OK) {
```
We usually go with just `if(ensure_trailing(&gateway, '/')) {` when checking CURLcode functions since we know `CURLE_OK` is zero. That saves us from comparing with `CURLE_OK` in checks. It's not a *must*, but it is a standard pattern in curl code. You could simplify several checks here like this.

markg85 replied:
This + comment below. Cleans it up nicely! Thanx :)

For reference. Some jobs back i was "forced" into writing checks verbose. The reasoning there was some people didn't like to have to pay attention to the single exclamation mark and wanted a policy where the `!` at the beginning of an if-statement wasn't allowed... That - now a habit of sorts - kinda stayed with me. Not that verbose is wrong, but short and still readable is better imho.

I like your style better :)

bagder commented on src/tool_operate.c:
```diff
@@ -808,7 +851,13 @@ static CURLcode ipfs_url_rewrite(CURLU *uh, const char *protocol, char **url,
+      goto clean;
+    }
+
+    if(curl_url_set(gatewaysurl, CURLUPART_URL, config->ipfs_gateway,
                     CURLU_GUESS_SCHEME)
                     == CURLUE_OK) {
```
*I* would write these conditions like this to make them slightly shorter and in my view easier to read:

```suggestion
    if(!curl_url_set(gatewaysurl, CURLUPART_URL, config->ipfs_gateway,
                     CURLU_GUESS_SCHEME)) {
```

bagder commented on src/tool_operate.c:
```diff
@@ -824,32 +873,92 @@ static CURLcode ipfs_url_rewrite(CURLU *uh, const char *protocol, char **url,
+  /* if the input path is just a slash, clear it */
+  if(inputpath && *inputpath && strlen(inputpath) == 1) {
+    if(*inputpath == '/') {
+      Curl_safefree(inputpath);
+      inputpath = strdup("");
+      if(!inputpath)
+        goto clean;
```
Since the memory is already allocated and a single byte that you want to not be there. Maybe just overwrite it?

```suggestion
      *inputpath = '\0';
```

markg85 commented:
Hi @bagder,

I just updated the patch with a little off-site review feedback. Only thing changed is the CID in the testcases (from `Qm...`, base58) to be of the "v1"  (`bafyt...` base32) format. It's only to discourage the `Qm...` format and not keep copying it over in new testcases.

Assuming it passes all the currently running tests, feel free to merge it!

mkauf commented:
There are failed tests for this change in the autobuilds. E. g. test 724: "Malformed URL". https://curl.se/dev/log.cgi?id=20231107030851-1844341#prob1 or https://curl.se/dev/log.cgi?id=20231107033909-1867421#prob1

bagder commented:
I did some further tidying up in 01d9b8bc7ceb56dc116 so it may things it a little.

markg85 commented:
I'm happy some testcases fail, means I've added enough cases to catch edits that might change the behavior :)

mkauf commented:
@markg85 / @bagder 

* There is an endianness bug. Proposed bugfix:
```
diff --git a/src/tool_ipfs.c b/src/tool_ipfs.c
index 271fff2cd..435d1697c 100644
--- a/src/tool_ipfs.c
+++ b/src/tool_ipfs.c
@@ -104,7 +104,8 @@ static char *ipfs_gateway(void)

     /* get the first line of the gateway file, ignore the rest */
     while((c = getc(gateway_file)) != EOF && c != '\n' && c != '\r') {
-      if(curlx_dyn_addn(&dyn, &c, 1))
+      char c_char = (char)c;
+      if(curlx_dyn_addn(&dyn, &c_char, 1))
         goto fail;
     }
```

* Some tests set the environment variable `IPFS_DATA`. But this variable is not checked at all in the curl tool code. Probably you wanted to set the environment variable `IPFS_PATH`.

* Typo in the tests: Change "traling" to "trailing"

* Wrong link in the documentation: Change "#Automatic-gateway-detection" to "#automatic-gateway-detection" in `docs/IPFS.md`

markg85 commented:
Hi @mkauf 

Thank you for that review, that's much appreciated!

> There is an endianness bug. Proposed bugfix:
I'll fix it with that suggestion.
Meanwhile, could you explain the reasoning why adding an extra temp variable with `c` casts as `char`  magically fixes endianess? Internally - in curlx_dyn_addn - the input is casted to `unsigned char *`. I'd assume that cast + the size of 1 (1 byte) would all be enough to catch endian differences, no?

> Some tests set the environment variable IPFS_DATA. But this variable is not checked at all in the curl tool code. Probably you wanted to set the environment variable IPFS_PATH.
Very good catch! While `IPFS_DATA` [does exist](https://docs.ipfs.tech/install/run-ipfs-inside-docker/), the intended use here is `IPFS_PATH`. Will fix the testcases.
The real thing that is being tested there though (tests 736, 737 and 738) isn't even the environment variable. It's making sure IPFS knows where to look for it's config file. Which by default is in $HOME/.ipfs. In other terms, tests 736 and 738 aren't testing properly and 737 just tests what's the default already.

> Typo in the tests: Change "traling" to "trailing"
Fixed.

> Wrong link in the documentation: Change "#Automatic-gateway-detection" to "#automatic-gateway-detection" in docs/IPFS.md
Fixed.

PR incoming.

mkauf commented:
@markg85

>  I'd assume that cast + the size of 1 (1 byte) would all be enough to catch endian differences, no?

No, because the pointer still points to an `int`. It still points to the same memory address as before.

These answers explain it in more detail:
* https://stackoverflow.com/a/21975205
* https://stackoverflow.com/a/21975272
